official/recommendation/data_download.py

```diff
@@ -0,0 +1,297 @@
+  np.random.seed(0)
+
+  # Load the file as DataFrame
+  file_path = os.path.join(data_dir, dataset_name)
+  df = load_file_to_df(file_path, sort=False)
+
+  # Get the info of users who have more than 20 ratings on items.
+  grouped = df.groupby(USER_COLUMN)
+  df = grouped.filter(lambda x: len(x) >= 20)
+  original_users = df[USER_COLUMN].unique()
+  original_items = df[ITEM_COLUMN].unique()
+
+  print('Generate user_map and item_map...')
+  user_map = {user: index for index, user in enumerate(original_users)}
+  item_map = {item: index for index, item in enumerate(original_items)}
+
+  df[USER_COLUMN] = df[USER_COLUMN].apply(lambda user: user_map[user])
+  df[ITEM_COLUMN] = df[ITEM_COLUMN].apply(lambda item: item_map[item])
+  assert df[USER_COLUMN].max() == len(original_users) - 1
+  assert df[ITEM_COLUMN].max() == len(original_items) - 1
+
+  # Generate data for train and test
+  all_ratings, test_ratings, test_negs = generate_train_eval_data(
+      df, original_users, original_items)
+
+  # serialize to csv file
+  df_train_ratings = pd.DataFrame(list(all_ratings))
+  df_train_ratings['fake_rating'] = 1
+  df_train_ratings.to_csv(
+      os.path.join(FLAGS.data_dir, dataset_name + '-' + TRAIN_RATINGS_FILENAME),
+      index=False, header=False, sep='\t')
+
+  df_test_ratings = pd.DataFrame(test_ratings)
+  df_test_ratings['fake_rating'] = 1
+  df_test_ratings.to_csv(
+      os.path.join(FLAGS.data_dir, dataset_name + '-' + TEST_RATINGS_FILENAME),
+      index=False, header=False, sep='\t')
+
+  df_test_negs = pd.DataFrame(test_negs)
+  df_test_negs.to_csv(
+      os.path.join(FLAGS.data_dir, dataset_name + '-' + TEST_NEG_FILENAME),
+      index=False, header=False, sep='\t')
+
+
+def main(_):
+  """Download and extract the data from grouplens website."""
+  if not os.path.exists(FLAGS.data_dir):
+    os.makedirs(FLAGS.data_dir)
+
+  # Download the dataset
+  dataset_zip = FLAGS.dataset + '.zip'
+  file_path = os.path.join(FLAGS.data_dir, dataset_zip)
+  if not os.path.exists(file_path):
+    def _progress(count, block_size, total_size):
+      sys.stdout.write('\r>> Downloading %s %.1f%%' % (
+          file_path, 100.0 * count * block_size / total_size))
+      sys.stdout.flush()
+
+    file_path, _ = urllib.request.urlretrieve(
+        DATA_URL + dataset_zip, file_path, _progress)
+    statinfo = os.stat(file_path)
+    print('\nSuccessfully downloaded', file_path, statinfo.st_size, 'bytes.')
+
+  # Unzip the dataset
+  if not os.path.exists(os.path.join(FLAGS.data_dir, FLAGS.dataset)):
+    zipfile.ZipFile(file_path, 'r').extractall(FLAGS.data_dir)
+
+  # Preprocess and parse the dataset to csv
+  if not os.path.exists(
+      os.path.join(
+          FLAGS.data_dir, FLAGS.dataset + '-' +TRAIN_RATINGS_FILENAME)):
+    parse_file_to_csv(FLAGS.data_dir, FLAGS.dataset)
+
+
+if __name__ == '__main__':
+  parser = argparse.ArgumentParser()
+  parser.add_argument(
+      '--data_dir', type=str, default='/tmp/ml_data',
```

Comment:
The default value "/tmp/ml_data" seems to be too generic. Can we give it a more specific name?

Reply:
Agreed. ml_data could also stand for "machine learning data", which could confuse the user.